plonky2/src/hash/field_merkle_tree.rs

```diff
@@ -0,0 +1,322 @@
+#[cfg(not(feature = "std"))]
+use alloc::vec;
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+use itertools::Itertools;
+
+use crate::hash::hash_types::RichField;
+use crate::hash::merkle_proofs::MerkleProof;
+use crate::hash::merkle_tree::{
+    capacity_up_to_mut, fill_digests_buf, merkle_tree_prove, MerkleCap,
+};
+use crate::plonk::config::{GenericHashOut, Hasher};
+use crate::util::log2_strict;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FieldMerkleTree<F: RichField, H: Hasher<F>> {
+    /// The data stored in the Merkle tree leaves.
+    pub leaves: Vec<Vec<Vec<F>>>,
+
+    /// Merkle tree node hashes, analogous to `digests` in `MerkleTree`.
+    pub digests: Vec<H::Hash>,
+
+    /// Represents the roots of the Merkle tree. This allows for using any layer as the root of the tree.
+    pub cap: MerkleCap<F, H>,
+
+    /// Represents the heights at which leaves reside within the tree.
+    pub leaf_heights: Vec<usize>,
+}
+
+impl<F: RichField, H: Hasher<F>> Default for FieldMerkleTree<F, H> {
+    fn default() -> Self {
+        Self {
+            leaves: Vec::new(),
+            digests: Vec::new(),
+            cap: MerkleCap::default(),
+            leaf_heights: Vec::new(),
+        }
+    }
+}
+
+impl<F: RichField, H: Hasher<F>> FieldMerkleTree<F, H> {
+    /// Each element in the `leaves` vector represents a matrix (a vector of vectors).
+    /// The height of each matrix should be a power of two.
+    /// The `leaves` vector should be sorted by matrix height, from tallest to shortest, with no duplicate heights.
```

Comment:
Btw, what do we do for duplicates?  Some kind of preprocessing?

Reply:
Added a TODO